packages/compass-aggregations/src/modules/aggregation.ts

```diff
@@ -39,32 +113,131 @@ export const runAggregation = (): ThunkAction<
   RootState,
   void,
   Actions
+> => {
+  return (dispatch) => dispatch(fetchAggregationData(1));
+};
+
+export const fetchPrevPage = (): ThunkAction<
+  void,
+  RootState,
+  void,
+  Actions
+> => {
+  return (dispatch, getState) => {
+    const {
+      aggregation: { page }
+    } = getState();
+    if (page <= 1) {
+      return;
+    }
+    dispatch(fetchAggregationData(page - 1));
+  };
+};
+
+export const fetchNextPage = (): ThunkAction<
+  void,
+  RootState,
+  void,
+  Actions
+> => {
+  return (dispatch, getState) => {
+    const {
+      aggregation: { isLast, page }
+    } = getState();
+    if (isLast) {
+      return;
+    }
+    dispatch(fetchAggregationData(page + 1));
+  };
+};
+
+export const cancelAggregation = (): ThunkAction<
+  void,
+  RootState,
+  void,
+  Actions
+> => {
+  return async (_dispatch, getState) => {
+    const {
+      dataService: { dataService },
+      aggregation: { abortController, session },
+    } = getState();
+    abortController?.abort();
+    if (session && dataService) {
+      try {
+        await dataService.killSessions(session);
+      } catch (e) {
+        log.warn(mongoLogId(1001000105), 'Aggregations', 'Attempting to kill the session failed');
+      }
+    }
+  };
+};
+
+const fetchAggregationData = (page: number): ThunkAction<
+  void,
+  RootState,
+  void,
+  Actions
 > => {
   return async (dispatch, getState) => {
     const {
       pipeline,
       namespace,
       maxTimeMS,
       collation,
-      dataService: { dataService }
+      dataService: { dataService },
+      aggregation: { limit },
     } = getState();
 
-    const stages = pipeline.map(generateStage);
-    const options: AggregateOptions = {
-      maxTimeMS: maxTimeMS || DEFAULT_MAX_TIME_MS,
-      allowDiskUse: true,
-      collation: collation || undefined,
-    };
-    if (dataService) {
-      const cursor = dataService.aggregate(
+    if (!dataService) {
+      return;
+    }
+
+    try {
+      const abortController = new AbortController();
+      const signal = abortController.signal;
+      const session = dataService.startSession('CRUD');
```

Comment:
I might be missing something, do we need both session and signal stored separately in the state for some reason? Session should be attached to the aggregate and canceled with it, so all this logic can be part of your `aggregatePipeline` method I think, I don't see why you would need to spread it across multiple actions

Reply:
moved session handling to aggregatePipeline :)